README.md
# loot_tables
Loot Table Simulator
Nicholas Lucas

The point of this porject is to provide a loot table builder and simulator all in one. You can use this to not only build loot tables (check out the loot-table-advanced library to learn more) but can be used to provide probabilistic distrubtions for completions of said loot tables.

Example:
Many modern video games use nested loot tables to hide unique boss items. Many are interested in obtaining a completed set of unique items at this boss. You can easily apply a distribution and find the average rate to obtain a certain unique item, but to complete an entire table of say 10 different unique items? 

Try and do the above and you will not go far. Thus this project was created to simulate, based on an input parameter, the average attempts needed complete a personalized loot table. The results are then plotted in a localhost app for your viewability. 

The average attempts or 'kc' required to complete a boss unique log is logged to the console, and you can find the most probably kc from the plot provided by the script.

More example scripts will be continuously added as the maintainer has time.

To run any of the scrips, go into the respective `.examples/game/` repo and run 
`node script.js` on the example you want, this one is named script.js.


CHANGELOG.md
### [1.0.1] - 04-24-2023

## Changed

- Fixed barrowsBrothers.js plot config.

### [1.0.0] - 04-24-2023

## Changed

- Initialization of loot table simulation project.

index.js
import {
    GetLoot,
    LootTableEntry,
  } from 'loot-table-advanced'
  
  function generate_loot() {
  
      /*const uniques = [
          LootTableEntry('head', 1),
          LootTableEntry('body', 1),
          LootTableEntry('legs', 1),
          LootTableEntry('weapon', 1)
      ]*/
  
      const dharok = [
          LootTableEntry('d_head', 1),
          LootTableEntry('d_body', 1),
          LootTableEntry('d_legs', 1),
          LootTableEntry('d_weapon', 1)
      ]
  
      const guthan = [
          LootTableEntry('g_head', 1),
          LootTableEntry('g_body', 1),
          LootTableEntry('g_legs', 1),
          LootTableEntry('g_weapon', 1)
      ]
  
      const karil = [
          LootTableEntry('k_head', 1),
          LootTableEntry('k_body', 1),
          LootTableEntry('k_legs', 1),
          LootTableEntry('k_weapon', 1)
      ]
  
      const ahrim = [
          LootTableEntry('a_head', 1),
          LootTableEntry('a_body', 1),
          LootTableEntry('a_legs', 1),
          LootTableEntry('a_weapon', 1)
      ]
  
      const torag = [
          LootTableEntry('t_head', 1),
          LootTableEntry('t_body', 1),
          LootTableEntry('t_legs', 1),
          LootTableEntry('t_weapon', 1)
      ]
  
      const verac = [
          LootTableEntry('v_head', 1),
          LootTableEntry('v_body', 1),
          LootTableEntry('v_legs', 1),
          LootTableEntry('v_weapon', 1)
      ]
  
      const bros = [
          LootTableEntry('@dharok', 1),
          LootTableEntry('@guthan', 1),
          LootTableEntry('@karil', 1),
          LootTableEntry('@ahrim', 1),
          LootTableEntry('@torag', 1),
          LootTableEntry('@verac', 1)
      ]
  
      const drops = [
          LootTableEntry('@bros', 1),
          LootTableEntry('normies', 14)
      ]
  
      function ResolveHelper(id) {
          switch (id) {
              case 'dharok':
                  return dharok;
              case 'ahrim':
                  return ahrim;
              case 'guthan':
                  return guthan;
              case 'karil':
                  return karil;
              case 'torag':
                  return torag;
              case 'verac':
                  return verac;
              case 'bros':
                  return bros;
              case 'drops':
                  return drops;
          }
          return null;
      }
  
      let d_head = 0
      let d_body = 0
      let d_legs = 0
      let d_weapon = 0
      let g_head = 0
      let g_body = 0
      let g_legs = 0 
      let g_weapon = 0
      let k_head = 0
      let k_body = 0
      let k_legs = 0
      let k_weapon = 0
      let t_head = 0
      let t_body = 0
      let t_legs = 0
      let t_weapon = 0
      let v_head = 0
      let v_body = 0
      let v_legs = 0
      let v_weapon = 0
      let a_head = 0
      let a_body = 0
      let a_legs = 0
      let a_weapon = 0
      let done = 10000
      let kc = 0
  
      for (let i=0; i<done; i++) {
          let loot = GetLoot(drops, 1, ResolveHelper)
          let id = loot[0].id
  
          switch (id) {
              case 'd_head':
                  d_head = d_head + 1
                  console.log('Dharok helm obtained')
                  break
              case 'd_body':
                  d_body = d_body + 1
                  console.log('Dharok body obtained')
                  break
              case 'd_legs':
                  d_legs = d_legs + 1
                  console.log('Dharok legs obtained')
                  break
              case 'd_weapon':
                  d_weapon = d_weapon + 1
                  console.log
              case 'g_head':
                  g_head = g_head + 1
                  console.log('Guthan helm obtained')
                  break
              case 'g_body':
                  g_body = g_body + 1
                  console.log('Guthan body obtained')
                  break
              case 'g_legs':
                  g_legs = g_legs + 1
                  console.log('Guthan legs obtained')
                  break
              case 'g_weapon':
                  g_weapon = g_weapon + 1
                  console.log('Guthan weapon obtained')
              case 'k_head':
                  k_head = k_head + 1
                  console.log('Karil helm obtained')
              case 'k_body':
                  k_body = k_body + 1
                  console.log('Karil body obtained')
              case 'k_legs':
                  k_legs = k_legs + 1
                  console.log('Karil legs obtained')
                  break
              case 'k_weapon':
                  k_weapon = k_weapon + 1
                  console.log('Karil weapon obtained')
                  break
              case 't_head':
                  t_head = t_head + 1
                  console.log('Torag helm obtained')
                  break
              case 't_body':
                  t_body = t_body + 1
                  console.log('Torag body obtained')
                  break
              case 't_legs':
                  t_legs = t_legs + 1
                  console.log('Torag legs obtained')
                  break
              case 't_weapon':
                  t_weapon = t_weapon + 1
                  console.log('Torag weapon obtained')
                  break
              case 'v_head':
                  v_head = v_head + 1
                  console.log('Verac helm obtained')
                  break
              case 'v_body':
                  v_body = v_body + 1
                  console.log('Verac body obtained')
                  break
              case 'v_legs':
                  v_legs = v_legs + 1
                  console.log('Verac legs obtained')
                  break
              case 'v_weapon':
                  v_weapon = v_weapon + 1
                  console.log('Verac weapon obtained')
                  break
              case 'a_head':
                  a_head = a_head + 1
                  console.log('Ahrim helm obtained')
                  break
              case 'a_body':
                  a_body = a_body + 1
                  console.log('Ahrim body obtained')
                  break
              case 'a_legs':
                  a_legs = a_legs + 1
                  console.log('Ahrim legs obtained')
                  break
              case 'a_weapon':
                  a_weapon = a_weapon + 1
                  console.log('Ahrim weapon obtained')
                  break
          }
          
          
          if (d_head >= 1 && d_body >= 1 && d_legs >= 1 && d_weapon >=1 && g_head >= 1 && g_body >= 1 && g_legs >= 1 && g_weapon >= 1 && k_head >= 1 && 
          k_body >= 1 && k_legs >= 1 && k_weapon >= 1 && t_head >= 1 && t_body >= 1 && t_legs >= 1 && t_weapon >= 1 && v_head >= 1 && v_body >= 1 && 
          v_legs >= 1 && v_weapon >= 1 && a_head >= 1 && a_body >= 1 && a_legs >= 1 && a_weapon >= 1) {
              console.log(`KC for completed log: '${i}'`)
              kc = i
              return kc
          }
  
          if (i === done-1) {
              console.log(`Dry log KC: '${done}'`)
              return done
          }
      }
  }
  
  function main() {
      let total_kc = 0
      let sims = 10000
      let avg_kc = 0
      for (let l=0; l<sims; l++){
          total_kc = total_kc + generate_loot() 
      }
      avg_kc = total_kc/sims
      console.log(`Average kc to complete log: '${avg_kc}'`)
  }
  
  main()
  

examples/osrs/tempoross.js
import {
    GetLoot,
    LootTableEntry,
} from 'loot-table-advanced'
import { plot } from 'nodeplotlib'
  
function generate_loot() {
    
    const drops = [
        LootTableEntry('pet', 1),
        LootTableEntry('tome', 5),
        LootTableEntry('fish', 5),
        LootTableEntry('box', 20),
        LootTableEntry('barrel', 20),
        LootTableEntry('normies', 7949)
    ]

    function ResolveHelper(id) {
        switch (id) {
            case 'drops':
                return drops;
        }
        return null;
    }

    let pet = 0
    let tome = 0
    let fish = 0
    let box = 0
    let barrel = 0
    let done = 30000
    let kc = 0

    for (let i=0; i<done; i++) {
        let loot = GetLoot(drops, 1, ResolveHelper)
        let id = loot[0].id

        switch (id) {
            case 'pet':
                pet = pet + 1
                if (pet === 1) {
                  console.log('Tiny tempor pet obtained')
                }
                break
            case 'tome':
                tome = tome + 1
                if (tome === 1) {
                  console.log('Tome of water obtained')
                }
                break
            case 'fish':
                fish = fish + 1
                if (fish === 1) {
                  console.log('Big harpoonfish obtained')
                }
                break
            case 'box':
                box = box + 1
                if (box === 1) {
                  console.log('Tackle box obtained')
                }
                break
            case 'barrel':
                barrel = barrel + 1
                if (barrel === 1) {
                  console.log('Fish barrel obtained')
                }
                break
        }
        
        
        if ( pet >= 1 && tome >= 1 && fish >= 1 && box >= 1 && barrel >= 1) {
            //console.log(`KC for completed log: '${i}'`)
            kc = i
            return kc
        }

        if (i === done-1) {
            //console.log(`Dry log KC: '${done}'`)
            return done
        }
    }
}

function main() {
    let total_kc = 0
    let sims = 10000
    let avg_kc = 0
    let kc_list = []

    for (let l=0; l<sims; l++){
        let temp_kc = generate_loot()
        total_kc = total_kc + temp_kc
        kc_list.push(temp_kc)
    }
    avg_kc = total_kc/sims

    const data = [
        {
          x: kc_list,
          type: 'histogram',
        },
      ]
      
    plot(data)

    console.log(`Average kc to complete Tempoross collection log: '${avg_kc}'`)
}

main()


examples/osrs/brimstoneChest.js
import {
    GetLoot,
    LootTableEntry,
} from 'loot-table-advanced'
import { plot } from 'nodeplotlib'
  
function generate_loot() {
    
    const drops = [
        LootTableEntry('mystic_hat', 1),
        LootTableEntry('mystic_top', 1),
        LootTableEntry('mystic_bottoms', 1),
        LootTableEntry('mystic_gloves', 1),
        LootTableEntry('mystic_boots', 1),
        LootTableEntry('broken_hasta', 5),
        LootTableEntry('normies', 990)
    ]

    function ResolveHelper(id) {
        switch (id) {
            case 'drops':
                return drops;
        }
        return null;
    }

    let mystic_hat = 0
    let mystic_top = 0
    let mystic_bottoms = 0
    let mystic_gloves = 0
    let mystic_boots = 0
    let broken_hasta = 0
    let done = 10000
    let kc = 0

    for (let i=0; i<done; i++) {
        let loot = GetLoot(drops, 1, ResolveHelper)
        let id = loot[0].id

        switch (id) {
            case 'mystic_hat':
                mystic_hat = mystic_hat + 1
                if (mystic_hat === 1) {
                  //console.log('Mystic hat dusk obtained')
                }
                break
            case 'mystic_top':
                mystic_top = mystic_top + 1
                if (mystic_top === 1) {
                  //console.log('Mystic top dusk obtained')
                }
                break
            case 'mystic_bottoms':
                mystic_bottoms = mystic_bottoms + 1
                if (mystic_bottoms === 1) {
                  //console.log('Mystic bottoms dusk obtained')
                }
                break
            case 'mystic_gloves':
                mystic_gloves = mystic_gloves + 1
                if (mystic_gloves === 1) {
                  //console.log('Mystic gloves dusk obtained')
                }
                break
            case 'mystic_boots':
                mystic_boots = mystic_boots + 1
                if (mystic_boots === 1) {
                  //console.log('Mystic boots dusk obtained')
                }
                break
            case 'broken_hasta':
                broken_hasta = broken_hasta + 1
                if (broken_hasta === 1) {
                  //console.log('Broken dragon hasta obtained')
                }
                break
        }
        
        
        if ( mystic_hat >= 1 && mystic_top >= 1 && mystic_bottoms >= 1 && mystic_boots >= 1 && mystic_gloves >= 1 && broken_hasta >= 1) {
            //console.log(`KC for completed log: '${i}'`)
            kc = i
            return kc
        }

        if (i === done-1) {
            //console.log(`Dry log KC: '${done}'`)
            return done
        }
    }
}

function main() {
    let total_kc = 0
    let sims = 10000
    let avg_kc = 0
    let kc_list = []

    for (let l=0; l<sims; l++){
        let temp_kc = generate_loot()
        total_kc = total_kc + temp_kc
        kc_list.push(temp_kc)
    }
    avg_kc = total_kc/sims

    const data = [
        {
          x: kc_list,
          type: 'histogram',
        },
      ]
      
    plot(data)

    console.log(`Average kc to complete log: '${avg_kc}'`)
}

main()


examples/osrs/barrowsBrothers.js
import {
    GetLoot,
    LootTableEntry,
  } from 'loot-table-advanced'
  import { plot } from 'nodeplotlib'
    
    function generate_loot() {
    
        /*const uniques = [
            LootTableEntry('head', 1),
            LootTableEntry('body', 1),
            LootTableEntry('legs', 1),
            LootTableEntry('weapon', 1)
        ]*/
    
        const dharok = [
            LootTableEntry('d_head', 1),
            LootTableEntry('d_body', 1),
            LootTableEntry('d_legs', 1),
            LootTableEntry('d_weapon', 1)
        ]
    
        const guthan = [
            LootTableEntry('g_head', 1),
            LootTableEntry('g_body', 1),
            LootTableEntry('g_legs', 1),
            LootTableEntry('g_weapon', 1)
        ]
    
        const karil = [
            LootTableEntry('k_head', 1),
            LootTableEntry('k_body', 1),
            LootTableEntry('k_legs', 1),
            LootTableEntry('k_weapon', 1)
        ]
    
        const ahrim = [
            LootTableEntry('a_head', 1),
            LootTableEntry('a_body', 1),
            LootTableEntry('a_legs', 1),
            LootTableEntry('a_weapon', 1)
        ]
    
        const torag = [
            LootTableEntry('t_head', 1),
            LootTableEntry('t_body', 1),
            LootTableEntry('t_legs', 1),
            LootTableEntry('t_weapon', 1)
        ]
    
        const verac = [
            LootTableEntry('v_head', 1),
            LootTableEntry('v_body', 1),
            LootTableEntry('v_legs', 1),
            LootTableEntry('v_weapon', 1)
        ]
    
        const bros = [
            LootTableEntry('@dharok', 1),
            LootTableEntry('@guthan', 1),
            LootTableEntry('@karil', 1),
            LootTableEntry('@ahrim', 1),
            LootTableEntry('@torag', 1),
            LootTableEntry('@verac', 1)
        ]
    
        const drops = [
            LootTableEntry('@bros', 1),
            LootTableEntry('normies', 14)
        ]
    
        function ResolveHelper(id) {
            switch (id) {
                case 'dharok':
                    return dharok;
                case 'ahrim':
                    return ahrim;
                case 'guthan':
                    return guthan;
                case 'karil':
                    return karil;
                case 'torag':
                    return torag;
                case 'verac':
                    return verac;
                case 'bros':
                    return bros;
                case 'drops':
                    return drops;
            }
            return null;
        }
    
        let d_head = 0
        let d_body = 0
        let d_legs = 0
        let d_weapon = 0
        let g_head = 0
        let g_body = 0
        let g_legs = 0 
        let g_weapon = 0
        let k_head = 0
        let k_body = 0
        let k_legs = 0
        let k_weapon = 0
        let t_head = 0
        let t_body = 0
        let t_legs = 0
        let t_weapon = 0
        let v_head = 0
        let v_body = 0
        let v_legs = 0
        let v_weapon = 0
        let a_head = 0
        let a_body = 0
        let a_legs = 0
        let a_weapon = 0
        let done = 10000
        let kc = 0
    
        for (let i=0; i<done; i++) {
            let loot = GetLoot(drops, 1, ResolveHelper)
            let id = loot[0].id
    
            switch (id) {
                case 'd_head':
                    d_head = d_head + 1
                    if (d_head === 1) {
                      console.log('Dharok helm obtained')
                    }
                    break
                case 'd_body':
                    d_body = d_body + 1
                    if (d_body === 1) {
                      console.log('Dharok body obtained')
                    }
                    break
                case 'd_legs':
                    d_legs = d_legs + 1
                    if(d_legs === 1) {
                      console.log('Dharok legs obtained')
                    }
                    break
                case 'd_weapon':
                    d_weapon = d_weapon + 1
                    if (d_weapon === 1) {
                      console.log('Dharok greataxe obtained')
                    }
                    break
                case 'g_head':
                    g_head = g_head + 1
                    if (g_head === 1) {
                      console.log('Guthan helm obtained')
                    }
                    break
                case 'g_body':
                    g_body = g_body + 1
                    if (g_body === 1) {
                      console.log('Guthan body obtained')
                    }
                    break
                case 'g_legs':
                    g_legs = g_legs + 1
                    if (g_legs === 1) {
                      console.log('Guthan legs obtained')
                    }
                    break
                case 'g_weapon':
                    g_weapon = g_weapon + 1
                    if (g_weapon === 1) {
                      console.log('Guthan spear obtained')
                    }
                    break
                case 'k_head':
                    k_head = k_head + 1
                    if (k_head === 1) {
                      console.log('Karil helm obtained')
                    }
                    break
                case 'k_body':
                    k_body = k_body + 1
                    if (k_body === 1) {
                      console.log('Karil body obtained')
                    }
                    break
                case 'k_legs':
                    k_legs = k_legs + 1
                    if (k_legs === 1) {
                      console.log('Karil legs obtained')
                    }
                    break
                case 'k_weapon':
                    k_weapon = k_weapon + 1
                    if (k_weapon === 1) {
                      console.log('Karil crossbow obtained')
                    }
                    break
                case 't_head':
                    t_head = t_head + 1
                    if (t_head === 1) {
                      console.log('Torag helm obtained')
                    }
                    break
                case 't_body':
                    t_body = t_body + 1
                    if (t_body === 1) {
                      console.log('Torag body obtained')
                    }
                    break
                case 't_legs':
                    t_legs = t_legs + 1
                    if (t_legs === 1) {
                      console.log('Torag legs obtained')
                    }
                    break
                case 't_weapon':
                    t_weapon = t_weapon + 1
                    if (t_weapon === 1) {
                      console.log('Torag hammers obtained')
                    }
                    break
                case 'v_head':
                    v_head = v_head + 1
                    if (v_head === 1) {
                      console.log('Verac helm obtained')
                    }
                    break
                case 'v_body':
                    v_body = v_body + 1
                    if (v_body === 1) {
                      console.log('Verac body obtained')
                    }
                    break
                case 'v_legs':
                    v_legs = v_legs + 1
                    if (v_legs === 1) {
                      console.log('Verac legs obtained')
                    }
                    break
                case 'v_weapon':
                    v_weapon = v_weapon + 1
                    if (v_weapon === 1) {
                      console.log('Verac flail obtained')
                    }
                    break
                case 'a_head':
                    a_head = a_head + 1
                    if (a_head === 1) {
                      console.log('Ahrim helm obtained')
                    }
                    break
                case 'a_body':
                    a_body = a_body + 1
                    if (a_body === 1) {
                      console.log('Ahrim legs obtained')
                    }
                    break
                case 'a_legs':
                    a_legs = a_legs + 1
                    if (a_legs === 1) {
                      console.log('Ahrim legs obtained')
                    }
                    break
                case 'a_weapon':
                    a_weapon = a_weapon + 1
                    if (a_weapon === 1) {
                      console.log('Ahrim staff obtained')
                    }
                    break
            }
            
            
            if (d_head >= 1 && d_body >= 1 && d_legs >= 1 && d_weapon >=1 && g_head >= 1 && g_body >= 1 && g_legs >= 1 && g_weapon >= 1 && k_head >= 1 && 
            k_body >= 1 && k_legs >= 1 && k_weapon >= 1 && t_head >= 1 && t_body >= 1 && t_legs >= 1 && t_weapon >= 1 && v_head >= 1 && v_body >= 1 && 
            v_legs >= 1 && v_weapon >= 1 && a_head >= 1 && a_body >= 1 && a_legs >= 1 && a_weapon >= 1) {
                console.log(`KC for completed log: '${i}'`)
                kc = i
                return kc
            }
    
            if (i === done-1) {
                console.log(`Dry log KC: '${done}'`)
                return done
            }
        }
    }
    
    function main() {
        let total_kc = 0
        let sims = 10000
        let avg_kc = 0
        let kc_list = []
  
        for (let l=0; l<sims; l++){
            let temp_kc = generate_loot()
            total_kc = total_kc + temp_kc
            kc_list.push(temp_kc)
        }
        avg_kc = total_kc/sims
        
        const data = [
          {
            x: kc_list,
            type: 'histogram',
          },
        ]
        
        plot(data)
  
        console.log(`Average kc to complete log: '${avg_kc}'`)
    }
    
    main()
    